packages/opentelemetry-scope-async-hooks/src/AsyncHooksScopeManager.ts

```diff
@@ -0,0 +1,158 @@
+import * as shimmer from 'shimmer';
+
+const EVENT_EMITTER_METHODS: Array<keyof EventEmitter> = [
+  'addListener',
+  'on',
+  'once',
+  'prependListener',
+  'prependOnceListener',
+];
+
+export class AsyncHooksScopeManager implements ScopeManager {
+  private asyncHook: asyncHooks.AsyncHook;
+  private scopes: { [uid: number]: unknown } = {};
+  private current: unknown = null;
+
+  constructor() {
+    this.asyncHook = asyncHooks.createHook({
+      init: this.init.bind(this),
+      before: this.before.bind(this),
+      destroy: this.destroy.bind(this),
+    });
+  }
+
+  active(): unknown {
+    return this.current;
+  }
+
+  with<T extends (...args: unknown[]) => ReturnType<T>>(
+    scope: unknown,
+    fn: T
+  ): ReturnType<T> {
+    const oldCurrent = this.current;
+    this.current = scope;
+    const res = fn();
+    this.current = oldCurrent;
+    return res;
+  }
+
+  bind<T>(target: T, scope?: unknown): T {
+    // if no specific scope to propagate is give, we use the current one
+    if (!scope) {
+      scope = this.current;
+    }
+    if (target instanceof EventEmitter) {
+      return this.bindEventEmitter(target, scope);
+    } else if (typeof target === 'function') {
+      return this.bindFunction(target, scope);
+    }
+    return target;
+  }
+
+  enable(): this {
+    this.asyncHook.enable();
+    return this;
+  }
+
+  disable(): this {
+    this.asyncHook.disable();
+    this.scopes = {};
+    this.current = null;
+    return this;
+  }
+
+  private bindFunction<T extends Function>(target: T, scope?: unknown): T {
+    const manager = this;
+    const boundContext = scope;
+    const contextWrapper = function(this: {}) {
+      const oldContext = manager.current;
+      manager.current = boundContext;
+      const res = target.apply(this, arguments);
+      manager.current = oldContext;
+      return res;
+    };
+    Object.defineProperty(contextWrapper, 'length', {
+      enumerable: false,
+      configurable: true,
+      writable: false,
+      value: target.length,
+    });
+    /**
+     * It isnt possible to tell Typescript that contextWrapper is the same as T
+     * so we forced to cast as any here.
+     */
+    // tslint:disable-next-line:no-any
+    return contextWrapper as any;
+  }
+
+  /**
+   * By default, EventEmitter call their callback with their scope, which we do
+   *  not want, instead we will bind a specific scope to all callbacks that
+   *  go through it.
+   * @param target EventEmitter a instance of EventEmitter to patch
+   * @param scope the scope we want to bind
+   */
+  private bindEventEmitter<T extends EventEmitter>(
+    target: T,
+    scope?: unknown
+  ): T {
+    const scopeManager = this;
+    EVENT_EMITTER_METHODS.forEach(methodName => {
+      if (target[methodName] === undefined) return;
+      shimmer.wrap(target as EventEmitter, methodName, (original: Function) => {
+        return function(this: {}, event: string, cb: Function) {
```

Comment:
If the same target is passed multiple times, the current implementation will keep wrapping it resulting in multiple wrappers and the wrong scope.

Reply:
The problem is that this implementation allows to bind a custom scope, so if we want to change the scope to another it wouldn't work, should this be documented as a limitation ?

Reply:
Will open another PR for that following discussion in SIG